[OpenCL] Check textures availability

pierrepaleo opened the pull request:
This PR adds a mechanism for checking textures (openCL "images") availability.

Depending on the underlying openCL implementation, we can never be sure that textures are actually available, before explicitly trying to create one.  
For example, with `pocl`:  `backprojector.ctx.devices[0].get_info(cl.device_info.IMAGE_SUPPORT)`  returns `1` while `backprojector.check_texture_availability()` returns `False`.

Close #2700, #3243

Changelog: OpenCL: add textures availability check

pierrepaleo commented:
Tested on the following machines.

## Machine `A`

Hardware: GeForce GTX 580, Quadro P400, Xeon Silver 4114 CPU  
Software: Debian 10, python 3.7

  - [x] Convolution: OK on all platforms. With CPU+POCL, we have warning about clean-up fails (dead contexts ?).
  - [x] Backprojection: OK on all platforms
  - [x] Projection: OK on all platforms

  ## Machine `B`

Hardware: Quadro P2000, Xeon E5-2643-0 CPU
Software: Ubuntu 20.04, python 3.8

  - [ ] Convolution: TODO
  - [ ] Backprojection: TODO
  - [ ] Projection: TODO


##  Machine `C`

Hardware: GTX 1080 Ti, Xeon CPU E5-2643 v3 
Software: Debian 8.8, python 3.7
OpenCL platforms:
```
[0] <pyopencl.Platform 'NVIDIA CUDA'>
[1] <pyopencl.Platform 'Intel(R) OpenCL'>
[2] <pyopencl.Platform 'AMD Accelerated Parallel Processing'>
```

  - [ ] Convolution: **FAIL**
  - [ ] Backprojection: **TODO** the first test fails on platforms 1 and 2
  - [x] Projection: OK on all platforms


## Machine `D`

Hardware: Tesla V100, Power9  
Software: Ubuntu 20.04, python 3.8

  - [ ] Convolution: TODO
  - [ ] Backprojection: TODO
  - [ ] Projection: TODO

pierrepaleo commented:
Funny thing: running
```bash
./run_tests.py silx.opencl.test.test_convolution.suite
```
works, but running
```bash
./run_tests.py silx.opencl.test.test_convolution
```
fails, because `unittest` does not seem to be able to pass the `param` argument at class initialization.

t20100 commented:
You can't provide a module as argument, it needs to be a `TestSuite`, a `TestCase` or a `TestCase` method.

pierrepaleo commented:
Understood, thanks !

kif commented:
Do you actually win anything by using textures ? how much (beside the approximate interpolator)?

pierrepaleo commented:
For convolution, on a GTX 1080 Ti:

```python
import numpy as np
from silx.opencl.convolution import Convolution
import pyopencl.array as parray
from scipy.misc import ascent
from silx.image.utils import gaussian_kernel
from time import sleep

img = ascent().astype("f")
img2 = np.tile(img, (8,8))
kern = gaussian_kernel(3)
C = Convolution((4096, 4096), kern, axes=(0, 1), extra_options={"dont_use_textures": False}, profile=True)
C0 = Convolution((4096, 4096), kern, axes=(0, 1), extra_options={"dont_use_textures": True}, profile=True)

d_img = parray.to_device(C.queue, img2)
d_o = parray.zeros_like(d_img)

def print_event(proc, i):
    evt = proc.events[i]
    t_ms = (evt.event.profile.end - evt.event.profile.start)/1e6
    print("%s: %.3f" % (evt.name, t_ms))

C(d_img, output=d_o)
sleep(1)
C(d_img, output=d_o)
sleep(1)
print_event(C, -1)
print_event(C, -2)
print_event(C, -3)

C0(d_img, output=d_o)
sleep(1)
C0(d_img, output=d_o)
sleep(1)
print_event(C0, -1)
print_event(C0, -2)
```

gives (timings in ms):

```
# textures
batched convolution: 2.637
Transfer to texture: 0.442
batched convolution: 2.661
# no textures
batched convolution: 10.590
batched convolution: 10.989
```
and this is for convolution, where no interpolation is done, only a 1D stencil pattern.  
A bit more work is to be done for comparing things with `Backprojector` (as it always use texture whenever available) but since the access pattern is a lot more complicated, I suspect the gain will be significant.

pierrepaleo commented:
Ready for review. 

Some items worth mentioning:
  - With POCL: `backprojector.ctx.devices[0].get_info(cl.device_info.IMAGE_SUPPORT)` returns `1` while `backprojector.check_texture_availability()` returns `False`  (this is one reason for using `check_texture_availability` in the first place).
  - Replaced `self.is_cpu` with `self._use_textures`. The two are not equivalent:
    - Textures might be available on CPU, depending on the OpenCL implementation, although certainly emulated. When using them, precision is lowered (so we still have to check whether the device is a CPU for the tests)
    - Textures might be not available on GPUs depending on the OpenCL implementation (ex. POCL)
  - With POCL + nvidia hardware: compilation fails when compiling a kernel with "image2d_" functions, so I added some `#ifndef` guards

kif commented:
I am reviewing it ...

kif commented:
is this branch on the silx-kit repo?

kif commented:
Looks good to me. I'll destroy the branch once merged.
